hub/util/merge.py

```diff
@@ -0,0 +1,321 @@
+    common_tensors = target_tensors & original_tensors
+
+    # present in dataset at lca, but deleted in target
+    target_deleted_tensors = lca_tensors - target_tensors
+
+    # present in dataset at lca, but deleted in original
+    original_deleted_tensors = lca_tensors - original_tensors
+
+    target_diff, _ = dataset.diff(target_commit_id, lca_id, as_dict=True)
+    for tensor in original_deleted_tensors:
+        diff = target_diff.get(tensor, None)
+
+        # either target doesn't have the tensor, no point in creating again
+        # or target has the tensor but it wasn't modified
+        if not diff or not (diff.data_added or diff.data_updated):
+            new_tensors.discard(tensor)
+
+    merge_common_tensors(
+        common_tensors,
+        dataset,
+        target_ds,
+        original_node,
+        target_node,
+        lca_node,
+        conflict_resolution,
+    )
+    copy_new_tensors(dataset, target_ds, new_tensors)
+
+    if delete_removed_tensors:
+        delete_tensors(dataset, target_deleted_tensors)
+
+    original_node.merge_from(target_node)
+    commit(dataset, f"Merge {target_id} into {dataset.branch}")
+
+
+def get_lca_tensors(ds, lca_id: str):
+    original_id = ds.pending_commit_id
+    ds.checkout(lca_id)
+    lca_tensors: Set[str] = set(ds.tensors.keys())
+    ds.checkout(original_id)
+    return lca_tensors
+
+
+def auto_commit_target_commit(dataset, target_commit_id: str):
+    original_id = dataset.pending_commit_id
+    original_branch = dataset.branch
+    dataset.checkout(target_commit_id)
+    auto_commit(dataset, f"Auto commit before merging into {original_branch}")
+    target_commit_id = dataset.pending_commit_id
+    dataset.checkout(original_id)
+    return target_commit_id
+
+
+def get_changes_commit_ids_for_node(
+    dataset, tensor_name: str, commit_node: Optional[CommitNode], lca_node: CommitNode
+):
+    changes_commit_map = defaultdict(list)
+    current_node = commit_node
+    while current_node and current_node.commit_id != lca_node.commit_id:
+        commit_id = current_node.commit_id
+        if current_node.is_merge_node:
+            changes = get_changes_commit_ids_for_node(
+                dataset, tensor_name, current_node.merge_parent, lca_node
+            )
+            for idx in changes:
+                changes_commit_map[idx].extend(changes[idx])
+        else:
+            try:
+                diff_key = get_tensor_commit_diff_key(tensor_name, commit_id)
+                diff: CommitDiff = dataset.storage.get_hub_object(diff_key, CommitDiff)
+                data_updated = sorted(diff.data_updated)
+                id_tensor_name = get_sample_id_tensor_key(tensor_name)
+                id_tensor = dataset[id_tensor_name]
+                for idx in data_updated:
+                    sample_id = id_tensor[idx].numpy()[0]
+                    changes_commit_map[sample_id].append(commit_id)
+            except KeyError:
```

Comment:
why key error might happen here?

Reply:
CommitDiff isn't present in some older hub versions when there are no changes made in a commit, for those commits we can skip as we know there are no changes

Reply:
better to explicitly check if it is older version (if commitDiff exists) rather than have a try/except covering 8 lines of code. 

It creates opportunity for bugs, also helps with readability.